File: LMJ/240615/PGS_이모티콘할인행사.java
import java.io.*;
import java.util.*;

class PGS_이모티콘할인행사 {
    static int N, ans, sum;
    static int[] discount, emoticons;
    static int[][] users;
    public int[] solution(int[][] u, int[] e) {
        N = e.length;
        ans = 0;
        sum = 0;
        discount = new int[N];
        users = u;
        emoticons = e;
        
        dfs(0);
        
        return new int[]{ans,sum};
    }
    
    static void dfs(int x){
        if(x==N){
            simulation();
            return;
        }
        for(int i = 10; i <= 40; i+=10){
            discount[x] = i;
            dfs(x+1);
        }
    }
    
    static void simulation(){
        int tmpAns = 0;
        int tmpSum = 0;
        for(int[] user: users){
            int tmp = 0;
            for(int i=0; i<N;i++){
                if(discount[i]>=user[0]){
                    tmp+=emoticons[i]*(100-discount[i])/100;
                }
            }
            if(tmp>=user[1]){
                tmpAns++;
            }
            else{
                tmpSum+=tmp;
            }
        }
        
        if(tmpAns>ans){
            ans = tmpAns;
            sum = tmpSum;
        } else if(tmpAns==ans && tmpSum > sum){
            sum = tmpSum;
        }
    }
}

File: HJH/240615/PGS_이모티콘 할인행사.py
from copy import deepcopy

candidate = []
disc = []
result = (0, 0)
u = 0
e = 0

def discount(money, percent):
    return int(money / 100 * (100 - percent))


def dfs(ith):
    global disc
    if ith == len(disc):
        global result, u, e
        money = [0] * len(u)
        sub = 0
        users = deepcopy(u)
        for di in range(len(disc)):
            for ui in range(len(users)):
                if users[ui][1] == 0:
                    continue
                if disc[di] >= users[ui][0]:
                    emo_disc = discount(e[di], disc[di])
                    if emo_disc >= users[ui][1]:
                        users[ui][1] = 0
                        money[ui] = 0
                        sub += 1
                    else:
                        users[ui][1] -= emo_disc
                        money[ui] += emo_disc
        if sub > result[0]:
            result = (sub, sum(money))
            # print(disc)
        elif sub == result[0] and sum(money) > result[1]:
            result = (sub, sum(money))
            # print(disc)

    else:
        global candidate
        for c in candidate:
            disc[ith] = c
            dfs(ith + 1)



def solution(users, emoticons):
    global disc, u, e, result, candidate
    users.sort()
    u = users
    e = emoticons
    disc = [0] * len(emoticons)
    c = set()
    for user in users:
        c.add(user[0])
    candidate = (10, 20, 30, 40)
    dfs(0)
    return list(result)